### PR TITLE
Demonstrating build interface

### LINKED ISSUE #331: build: Adjust `run_text()` to deal with builder output assumptions
The builder terminates in one of four states, depending on whether the DockerRequirement is specified as a "hint" or a "Requirement":

1. The builder has satisfied environment setup concerns and returned a subprocess object with command inputs and outputs.
2. The builder failed to satisfy environment setup concerns and returned "1" if Requirement, "0" if Hint.
3. The builder already satisfied the environment setup, found a container tarball in the "`build_cache`", and returns "0"
4. The builder inspected CWL hints and requirements for this task, no container was specified, and returns "0".

The `run_text()` function in the container runtime should assume that the environment is already resolved. If the builder failed in a Requirement, then the workflow should error and stop executing. Otherwise, the run_text can assume a container exists in the `build_cache`. If a container does not exist in the build cache, it should attempt to run baremetal. If it fails, report that and stop workflow execution.

I will consider this issue as part of #327 .

### PR DESCRIPTION
**Pull request by qwofford**
Tracking progress toward issue #326. Currently have a draft of how we might integrate the build_interface into the task_manager. See diff for details, feel free to discuss here or in weekly meeting.

@Boogie3D I'm now at a point where using `BuildTask` instead of a normal `Task` is a blocker. Perhaps there's something I can do to help or we can brainstorm on ways to kinda hack things to work?

### CODE REVIEW

**Comment by qwofford** [comment]
I finished the build interface integration draft. What do I mean by that? In order for the build system to be useful it has to integrate with the run system (`crt_drivers.py`). This means I had to change substantial parts of the crt. I do not expect this first iteration to be perfect, but it is a working example we can improve. Here's some sample output from the crt logs:
```
(beeflow-dedPeNMd-py3.8) -bash-4.2$ tail -f ~/.beeflow/logs/crt_driver.log 
INFO: Added clamr task to the submit queue
INFO: Resolving environment for task clamr
INFO: Environment preparation complete for task clamr
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: No container name provided. Assuming copyContainer source is runtime target.
INFO: Moving with the expectation that clamr-toss is the runtime container target
INFO: Expecting container at /yellow/users/qwofford/.beeflow/container_archive/clamr-toss.tar.gz. Ready to deploy and run.
INFO: Job Submitted clamr: job_id: 421362 job_state: PENDING
INFO: clamr PENDING -> RUNNING
INFO: clamr RUNNING -> COMPLETED
INFO: Added ffmpeg task to the submit queue
INFO: Job 421362 done clamr: removed from job status queue
INFO: Resolving environment for task ffmpeg
INFO: Environment preparation complete for task ffmpeg
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: No container name provided. Assuming copyContainer source is runtime target.
WARNING: No container specified, and cannot be inferred from copyContainer.
WARNING: Maybe you do not need a container, or try adding containerName or copyContainer.
INFO: Job Submitted ffmpeg: job_id: 421363 job_state: PENDING
INFO: ffmpeg PENDING -> COMPLETING
INFO: ffmpeg COMPLETING -> COMPLETED
INFO: Job 421363 done ffmpeg: removed from job status queue
```

And as a reminder, this is what the build interface output looks like:
```
(beeflow-dedPeNMd-py3.8) -bash-4.2$ cat ~/.beeflow/logs/builder.log 
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Deployed image root directory is: /var/tmp/qwofford/beeflow
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Container-relative output path is: /
INFO: task 24a8c427-95fb-44d9-a202-b3668dd3a7ba requirements has no DockerRequirements
INFO: task 24a8c427-95fb-44d9-a202-b3668dd3a7ba hint DockerRequirements: {'copyContainer'}
INFO: task 24a8c427-95fb-44d9-a202-b3668dd3a7ba union DockerRequirements consist of: {'copyContainer'}
INFO: task 24a8c427-95fb-44d9-a202-b3668dd3a7ba DockerRequirement execution order will be: ['copyContainer']
INFO: Execution order pre-empts hint/requirement status.
INFO: CharliecloudBuildDriver initialized
INFO: Executing build operation: "copyContainer"
INFO: Build will copy a container to /yellow/users/qwofford/.beeflow/container_archive/clamr-toss.tar.gz
INFO: Copying container.
INFO: Executing: cp /usr/projects/beedev/clamr/clamr-toss.tar.gz /yellow/users/qwofford/.beeflow/container_archive/clamr-toss.tar.gz

INFO: Reached terminal build case
INFO: Out of build instructions. Build operations complete.
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Deployed image root directory is: /var/tmp/qwofford/beeflow
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Container-relative output path is: /
INFO: task adbf0511-72df-4b33-a031-11c1ee005879 requirements has no DockerRequirements
INFO: task adbf0511-72df-4b33-a031-11c1ee005879 hints has no DockerRequirements
INFO: task adbf0511-72df-4b33-a031-11c1ee005879 union DockerRequirements consist of: set()
INFO: task adbf0511-72df-4b33-a031-11c1ee005879 DockerRequirement execution order will be: []
INFO: Execution order pre-empts hint/requirement status.
INFO: CharliecloudBuildDriver initialized
INFO: No build instructions provided. Assuming this is ok...
INFO: Out of build instructions. Build operations complete.
```

There are a ton of corner cases captured in what I've done so far, and a ton more that I'm sure I didn't think of. We will have to discuss them to know how the integration might need to change.

**Comment by qwofford** [comment]
Oh yeah, I also added the "copyContainer" command to supplement the CWL spec. This covers the case where we have a tarball we want to copy into BEE's build cache. We might consider merging the build cache with rusty's archive...

**Comment by qwofford** [comment]
Added a test case that includes both copyContainer and containerName:
```
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Deployed image root directory is: /var/tmp/qwofford/beeflow
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Container-relative output path is: /
INFO: task 08d8d6c9-c63d-48bb-a079-db7d0ca98348 requirements has no DockerRequirements
INFO: task 08d8d6c9-c63d-48bb-a079-db7d0ca98348 hint DockerRequirements: {'copyContainer', 'containerName'}
INFO: task 08d8d6c9-c63d-48bb-a079-db7d0ca98348 union DockerRequirements consist of: {'copyContainer', 'containerName'}
INFO: task 08d8d6c9-c63d-48bb-a079-db7d0ca98348 DockerRequirement execution order will be: ['containerName', 'copyContainer']
INFO: Execution order pre-empts hint/requirement status.
INFO: CharliecloudBuildDriver initialized
INFO: Executing build operation: "containerName"
INFO: Executing build operation: "copyContainer"
INFO: Build will copy a container to /yellow/users/qwofford/.beeflow/container_archive/foo.tar.gz
INFO: Copying container.
INFO: Executing: cp /usr/projects/beedev/clamr/clamr-toss.tar.gz /yellow/users/qwofford/.beeflow/container_archive/foo.tar.gz

INFO: Reached terminal build case
INFO: Out of build instructions. Build operations complete.
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Deployed image root directory is: /var/tmp/qwofford/beeflow
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Container-relative output path is: /
INFO: task c54ca389-54fa-424c-90a1-0022f243b35f requirements has no DockerRequirements
INFO: task c54ca389-54fa-424c-90a1-0022f243b35f hints has no DockerRequirements
INFO: task c54ca389-54fa-424c-90a1-0022f243b35f union DockerRequirements consist of: set()
INFO: task c54ca389-54fa-424c-90a1-0022f243b35f DockerRequirement execution order will be: []
INFO: Execution order pre-empts hint/requirement status.
INFO: CharliecloudBuildDriver initialized
INFO: No build instructions provided. Assuming this is ok...
INFO: Out of build instructions. Build operations complete.
```

**Comment by qwofford** [comment]
In previous discussions, we decided to stick with Docker's interpretation of "dockerImageId" rather than the intepretation CWL uses. I modified the builder tool to account for this. I also added a bee workflow that demonstrates this behavior:
```
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Deployed image root directory is: /var/tmp/qwofford/beeflow
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Container-relative output path is: /
INFO: task b6d77018-f4fd-4099-a7cd-101a9217a574 requirements has no DockerRequirements
INFO: task b6d77018-f4fd-4099-a7cd-101a9217a574 hint DockerRequirements: {'containerName', 'dockerFile'}
INFO: task b6d77018-f4fd-4099-a7cd-101a9217a574 union DockerRequirements consist of: {'containerName', 'dockerFile'}
INFO: task b6d77018-f4fd-4099-a7cd-101a9217a574 DockerRequirement execution order will be: ['containerName', 'dockerFile']
INFO: Execution order pre-empts hint/requirement status.
INFO: CharliecloudBuildDriver initialized
INFO: Executing build operation: "containerName"
INFO: Executing build operation: "dockerFile"
INFO: Context directory will be /tmp/foo2.
INFO: Build will create tar ball at /yellow/users/qwofford/.beeflow/container_archive/foo2.tar.gz
INFO: Context directory configured. Beginning build.
INFO: Executing: ch-image build -t foo2 -f src/beeflow/data/dockerfiles/Dockerfile.clamr-lanl-x86_64 /tmp/foo2
ch-builder2tar foo2 /yellow/users/qwofford/.beeflow/container_archive
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Deployed image root directory is: /var/tmp/qwofford/beeflow
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Container-relative output path is: /
INFO: task 241acdd2-84aa-4e97-b91e-6b14c63fd202 requirements has no DockerRequirements
INFO: task 241acdd2-84aa-4e97-b91e-6b14c63fd202 hints has no DockerRequirements
INFO: task 241acdd2-84aa-4e97-b91e-6b14c63fd202 union DockerRequirements consist of: set()
INFO: task 241acdd2-84aa-4e97-b91e-6b14c63fd202 DockerRequirement execution order will be: []
INFO: Execution order pre-empts hint/requirement status.
INFO: CharliecloudBuildDriver initialized
INFO: No build instructions provided. Assuming this is ok...
INFO: Out of build instructions. Build operations complete.
INFO: Reached terminal build case
INFO: Out of build instructions. Build operations complete.
```

**Comment by qwofford** [comment]
Updated documentation and modified container_drivers to handle dockerFile builds/deploys using the containerName convention rather than the dockerImageId convention:
```
>>> from beeflow.common.build.container_drivers import CharliecloudBuildDriver
>>> from beeflow.common.wf_data import Task
>>> task = Task(name='hi',command=['hi','hello'],
...                  requirements={'DockerRequirement':{'dockerFile':'src/beeflow/data/dockerfiles/Dockerfile.builder_demo',
...                                                     'containerName':'my_fun_container:sillytag'}},
...                  hints=None,
...                  workflow_id=42,
...                  subworkflow=None,
...                  inputs={},
...                  outputs={})
>>> b = CharliecloudBuildDriver(task)
INFO: Build container archive directory is: /yellow/users/qwofford/.beeflow/container_archive
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Deployed image root directory is: /var/tmp/qwofford/beeflow
INFO: Wrote deployed image root to user BeeConfig file.
INFO: Container-relative output path is: /
INFO: task e08dd293-6093-4882-910a-2463657c679d requirement DockerRequirements: {'containerName', 'dockerFile'}
INFO: task e08dd293-6093-4882-910a-2463657c679d hints has no DockerRequirements
INFO: task e08dd293-6093-4882-910a-2463657c679d union DockerRequirements consist of: {'containerName', 'dockerFile'}
INFO: task e08dd293-6093-4882-910a-2463657c679d DockerRequirement execution order will be: ['containerName', 'dockerFile']
INFO: Execution order pre-empts hint/requirement status.
>>> b.containerName()
0
>>> b.dockerFile()
INFO: Context directory will be /tmp/my_fun_container:sillytag.
INFO: Build will create tar ball at /yellow/users/qwofford/.beeflow/container_archive/my_fun_container:sillytag.tar.gz
INFO: Context directory configured. Beginning build.
INFO: Executing: ch-image build -t my_fun_container:sillytag -f src/beeflow/data/dockerfiles/Dockerfile.builder_demo /tmp/my_fun_container:sillytag
ch-builder2tar my_fun_container:sillytag /yellow/users/qwofford/.beeflow/container_archive
CompletedProcess(args='ch-image build -t my_fun_container:sillytag -f src/beeflow/data/dockerfiles/Dockerfile.builder_demo /tmp/my_fun_container:sillytag\nch-builder2tar my_fun_container:sillytag /yellow/users/qwofford/.beeflow/container_archive', returncode=0, stdout=b'builder: ch-image\nexporting\n/var/tmp/qwofford/ch-image/img/my_fun_container:sillytag\ncompressing\n-rw-rw---- 1 qwofford viz_public 72M Jul 28 10:06 /yellow/users/qwofford/.beeflow/container_archive/my_fun_container:sillytag.tar.gz\n', stderr=b'  1 FROM git.lanl.gov:5050/trandles/baseimages/centos:7\nmanifest: downloading\nconfig: downloading\nlayer 1/1: 524b0c1: downloading\nflattening image\nlayer 1/1: 524b0c1: listing\nvalidating tarball members\nresolving whiteouts\nlayer 1/1: 524b0c1: extracting\navailable --force: rhel7: CentOS/RHEL 7\nwarning: not yet supported, ignored: issue #780: CMD instruction\ngrown in 1 instructions: my_fun_container:sillytag\n')
(beeflow-dedPeNMd-py3.8) -bash-4.2$ ls -altrh ~/.beeflow/container_archive/
total 434M
drwxrwx--- 7 qwofford viz_public 4.0K Jul 26 08:19 ..
-rw-rw---- 1 qwofford viz_public 360M Jul 26 08:19 git.lanl.gov:5050%qwofford%containerhub%clamr-lanl-x86_64:latest.tar.gz
-rw-rw---- 1 qwofford viz_public  72M Jul 28 10:06 my_fun_container:sillytag.tar.gz
drwxrwx--- 2 qwofford viz_public 4.0K Jul 28 10:06 .
```

**Comment by qwofford** [comment]
Note that I removed the crt image_ok check from `crt_driver` and I removed references to that check from the slurm worker and worker interfaces. The reasons I did this are:
1. Checking on the status of a runtime environment is now the responsibility of the builder. I believe the CRT should assume the environment is prepared, because the build pre-task model handles that.
2. It is not necessarily a problem if the compute node doesn't have access to a container. It may be a baremetal environment step.

**Comment by pagrubel** [comment]
Sounds good and thanks for letting us know why you removed it.

**Comment by qwofford** [comment]
Making some notes here, this is challenging to troubleshoot:
1. Task manager node fails to build a valid container during the build pre-task. If I copy and paste the build commands interactively, it builds fine. Investigate python subprocess problems. Instead, it builds and archives a tarball that is 20B in size, so it _looks_ like the container is valid when it's not.
2. When the build pre-task concludes, a tarball exists in the container archive (though incorrect). Next, an sbatch script is launched and the compute note can not see the container tarball, I suspect because NFS caching hasn't finished setting up. However, you can create the container before hand, re-run the builder, and expect this tarball to be picked up by the container runtime on the compute node (because it was previously created with a large enough time delta)
3. Simply running the workflow twice solves the (supposed) NFS caching problem, but does not solve the build problem. You can, however, copy-paste the build command into the terminal to generate a valid container in the container_archive, _then_ re-run the workflow.
4. The Dockerfile for our CLAMR workflow is not exactly the same as the one we were using for testing in `/usr/projects/beeflow/`, and so the path being referenced in the previous CWL file does not match the path in the new container, which is defined in `src/beeflow/data/dockerfiles/Dockerfile.clamr-lanl-x86_64`.


I am confident that all of these problems can be solved, just wanted to list them all out explicitly to jog my memory as troubleshooting progresses.

**Comment by qwofford** [comment]
Toward 4. described above:
* I validated that the CLAMR container had a different path to the `clamr_cpuonly` binary inside the container than was referenced in the CWL file.
* I updated the CWL file in this repository and tested it. The CLAMR movie was successfully generated.
* Issue 4 is resolved, but issues 1-3 need to be addressed.

**Comment by qwofford** [comment]
Toward 1. described above:
* It appears that the build error is no longer appearing consistently. 
* I have noticed that /var/tmp is not big enough to contain both the ch-image directory for the CLAMR container, _and_ the deployed container which BEE stores by default in `/var/tmp/username/beeflow_deployed_containers`. This does not explain the issue, because the container is built on the front and, but deployed on a compute node. The compute node should have a blank `/var/tmp`.
* Because this issue is no longer reproducible, I will move on to troubleshooting 2.

**Comment by qwofford** [comment]
Jake found the bug! The issue was not related to NFS caching, but rather a non-blocking Python subprocess call. Once this was resolved the build interface worked for all 4 sample CLAMR workflows, which together represent the core functionality of the build interface.

Next step is to bring this branch up to date with develop.

**Comment by qwofford** [comment]
Ready for merge into develop as of today, but happy to revise once Steven's branch is merged in.

**Comment by pagrubel** [comment]
The examples in src/beeflow/build/README.md may need some updating, or I may need some instructions.

The clamr workflow cf.cwl for fog works, but there are some things in the task_manager log that we should probably delete. One is an error that turns out to be in develop. This note is just to remind me.

I need to check out some more of the build functionalities.

**Comment by pagrubel** [comment]
oh the log error or state error was not in develop but in this branch so noting it here:

INFO: Job Submitted clamr: job_id: 422586 job_state: PENDING
ERROR: Task Manager submit task clamr failed! 
 'str' object has no attribute 'class_'
ERROR: clamr state: SUBMIT_FAIL
INFO: clamr PENDING -> RUNNING

**Comment by pagrubel** [comment]
It seems there is too much in the try block in task_manager starting at line 159.

**Comment by qwofford** [comment]
@pagrubel Can I interpret this comment to be about Pylama code checking more generally? I'll turn on The Llama and start correcting for compliance.

**Comment by pagrubel** [comment]
Actually I think that particular block of code is cause the error I mentioned because something is not checked if I recall

**Comment by qwofford** [comment]
Marking WIP until pylama compliance is in order...

**Comment by pagrubel** [comment]
@qwofford so the gen_task_metadata was sending job_id, hostname and hash of the container to the gdb for provenance
not sure how you want to handle that

**Comment by qwofford** [comment]
Perhaps that is a task_manager issue to be handled in another PR. Now that the build tool is more general, task_manager will need to be extended too. See https://github.com/lanl/BEE_Private/issues/360

**Comment by pagrubel** [comment]
Fair enough!

**Comment by qwofford** [comment]
I believe the build interface integration is ready to merge into develop.

**Comment by pagrubel** [comment]
I tested the PR with the all the cwl files in src/beeflow/data/cwl/bee_workflows/clamr-ffmpeg/fog/no_yaml as well as failure due to using a directory in /tmp that I don't have access to. That fails and there is a warning in the appropriate logs; the steps fail. 

 The only problem I found was with trying to use requirements instead of hints. I'm wondering if the fix to requirements didn't get into the version of the parser that is being used in this PR. I did some debugging and the requirements just don't get passed into the task object.

File: src/beeflow/common/build/container_drivers.py [hunk 1]
```diff
@@ -302,14 +284,14 @@ def dockerFile(self, task_imageid=None, task_dockerfile=None, force=False):
             log.error("dockerFile not specified as task attribute or parameter.")
             return 1
 
-        # Create context directory to use as Dockerfile context, use task ID so user
+        # Create context directory to use as Dockerfile context, use container name so user
         # can prep the directory with COPY sources as needed.
-        context_dir = '/'.join(['/tmp',task_imageid])
+        context_dir = '/'.join(['/tmp', self.container_name])
```
Comment (pagrubel) [inline src/beeflow/common/build/container_drivers.py]:
```suggestion
        context_dir = '/'.join(['/tmp', getpass.getuser(), self.container_name])
```
```suggestion
        context_dir = '/'.join(['/tmp', self.container_name])
```

Comment (pagrubel) [reply src/beeflow/common/build/container_drivers.py]: After discussion with qwofford setting context_dir to the users home directory, will submit soon.

**Comment by pagrubel** [comment]
@qwofford So if an docker operation doesn't work you just give a warning?
Is this true for all options. It looks that way to me in the code and I tried it with pulling while I had CH_IMAGE_STORAGE=/tmp/invalid_bee_cache/ch-image
This is in the log
Executing build operation: "dockerPull"
WARNING: There was a problem executing dockerPull, check relevant log for detail.
INFO: Out of build instructions. Build operations complete.

I just want to make sure I understand what is going on. The clamr step looks as if it ran and ffmpeg fails because clamr step could not find the image. I'm noting this here although it is probably something we need to look at separately. I guess my main question is:
Should it just be a warning if you can't pull the container image?

**Comment by qwofford** [comment]
I think the argument for it being a warning is that the workflow may succeed even if the container pull fails. If the container were previously stored in the archive, the runner might use it and produce the desired research artifact despite the dockerPull failure.

When I implemented this as a warning, my thought was that you could specify a series of things to try and let them fail until one succeeds. The priority is set such that the fastest things are tried first, with a Dockerfile build being the option of last resort. This could be a time saving feature. For example, if a dockerPull and a dockerFile are specified, the dockerPull will try to pull, if it fails, it will try a dockerFile. The dockerFile build will take longer than a pull, but if the pull fails the job doesn't explicitly require that a pull succeeds, it requires that an environment is available.

**Comment by qwofford** [comment]
I'm not sure that feature has been tested, but if it doesn't work as intended I think that's a bugfix rather than a blocker. I think blocking issues are ones that directly conflict with what's working in develop, right?

**Comment by pagrubel** [comment]
I agree it is not a blocker. I just wanted to understand the path.

**Review by pagrubel — APPROVED** [review]

**Pull request merged**